spring-native-configuration/src/main/java/org/springframework/ValidatedComponentProcessor.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2019-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework;
+
+import org.springframework.nativex.hint.ProxyBits;
+import org.springframework.nativex.type.ComponentProcessor;
+import org.springframework.nativex.type.NativeContext;
+import org.springframework.nativex.type.Type;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Recognize spring.components that need validate proxies and register them.
+ *
+ * @author Petr Hejl
+ */
+public class ValidatedComponentProcessor implements ComponentProcessor {
+
+	@Override
+	public boolean handle(NativeContext imageContext, String componentType, List<String> classifiers) {
+		Type type = imageContext.getTypeSystem().resolveName(componentType);
+		return type.isAtValidated(true);
+	}
+
+	@Override
+	public void process(NativeContext imageContext, String componentType, List<String> classifiers) {
+		Type type = imageContext.getTypeSystem().resolveName(componentType);
+		List<String> transactionalInterfaces = new ArrayList<>();
+		boolean hasInterfaceMethods = false;
+		for (Type intface: type.getAllInterfaces()) {
+			transactionalInterfaces.add(intface.getDottedName());
+			if (!intface.getMethods().isEmpty()) {
+				hasInterfaceMethods = true;
+			}
+		}
+		if (!transactionalInterfaces.isEmpty() && hasInterfaceMethods) {
+			transactionalInterfaces.add("org.springframework.aop.SpringProxy");
+			transactionalInterfaces.add("org.springframework.aop.framework.Advised");
+			transactionalInterfaces.add("org.springframework.core.DecoratingProxy");
+			imageContext.addProxy(transactionalInterfaces);
+			imageContext.log(ValidatedComponentProcessor.class.getSimpleName() + ": creating proxy for these interfaces: " + transactionalInterfaces);
+		} else if (!type.isInterface()) {
```

Comment:
We should push this check up - there is no need to gather and walk through the interfaces if the type cannot be JDKProxied. (And this should be fixed in the transactional comp processor too).

Reply:
I'm not sure about this as we first check for interfaces (and implemented methods) and only otherwise (else) if it is a class we create aot proxy. Am I missing something?

Reply:
Before we had the ability to proxy classes we just 'had a go' at creating a JDKProxy for the interfaces a class implemented. This wasn't really correct but we got away with it sometimes. We added class proxies support and shoved them in as an afterthought, at least on the transactional component processor (in that processor there is still digging to do there as different samples seem to be satisfied with one vs the other - see how it isn't `else if` in the transactional processor, it will sometimes create both!).   I think this is really why I wanted to see some changes to the real sample around validation - add a few variant types with validation expressed in different ways (on the class itself, on interfaces, etc) - that's how we can truly discover if the component processor is doing the job it needs too.

Reply:
(If we aren't careful we can proxy too much - and it may work but images will be unnecessarily bloated with stuff that wasn't required)

Reply:
I agree. Once this is done I would like to check the transactional case and whether similar rules apply.